src/network/connectionpool.py

```diff
@@ -297,6 +298,17 @@ def loop(self):  # pylint: disable=too-many-branches,too-many-statements
                     # don't connect to self
                     if chosen in state.ownAddresses:
                         continue
+                    # don't connect to the hosts from the same
+                    # network group, defense against sibyl attacks
+                    host_network_group = protocol.network_group(
+                        chosen.host)
+                    same_group = False
+                    for j in self.outboundConnections.values():
+                        if host_network_group == j.network_group:
+                            same_group = True
+                            break
+                    if same_group:
```

Comment:
Maybe add `knownnodes.decreaseRating(chosen)`?

Reply:
The thing is, it may not be an attack, it could simply be that nodes run by different people are in the same data center. I don't think it's a sufficient reason to decrease rating.

Reply:
OK, maybe
```patch
diff --git a/src/network/connectionpool.py b/src/network/connectionpool.py
index 64533d9b..6264191d 100644
--- a/src/network/connectionpool.py
+++ b/src/network/connectionpool.py
@@ -306,6 +306,8 @@ class BMConnectionPool(object):
                     for j in self.outboundConnections.values():
                         if host_network_group == j.network_group:
                             same_group = True
+                            if chosen.host == j.destination.host:
+                                knownnodes.decreaseRating(chosen)
                             break
                     if same_group:
                         continue
```

Reply:
That's a reasonable suggestion, I'll do that.